File: arduinobot_ws/install/arduinobot_utils/local/lib/python3.10/dist-packages/arduinobot_utils/angle_conversion.py
#!/usr/bin/env python3

import rclpy  # Importa la librería principal de ROS 2 para Python
from rclpy.node import Node  # Importa la clase base para crear nodos en ROS 2
from arduinobot_msgs.srv import EulerToQuaternion, QuaternionToEuler
from tf_transformations import quaternion_from_euler, euler_from_quaternion

class AnglesConverter(Node):  # Define un nodo que expone servicios de conversión de ángulos
    def __init__(self):  # Constructor del nodo
        super().__init__("angles_converter")  # Inicializa el nodo con el nombre "angles_converter"
        
        self.euler_to_quaternion_ = self.create_service(  # Crea el servidor del servicio y lo registra en ROS 2
            EulerToQuaternion,                      # Tipo del servicio
            "euler_to_quaternion",                  # Nombre del servicio en el grafo de ROS 2
            self.euelerToQuaternionCallback         # Función callback que atiende las peticiones del servicio
        )

        self.quaternion_to_euler_ = self.create_service(  # Crea el servidor del servicio y lo registra en ROS 2
            QuaternionToEuler,                      # Tipo del servicio
            "quaternion_to_euler",                  # Nombre del servicio en el grafo de ROS 2
            self.quaternionToEulerCallback          # Función callback que atiende las peticiones del servicio
        )
        self.get_logger().info("Angle Conversion Services are ready")  # Mensaje de log indicando que los servicios están listos

    def euelerToQuaternionCallback(self, req, res):  # Callback del servicio: recibe la petición (req) y escribe la respuesta (res)
        self.get_logger().info("Requested to convert euler angles roll: %f, pitch: %f, yaw: %f, into a quaternion." % (req.roll, req.pitch, req.yaw))
        (res.x, res.y, res.z, res.w) = quaternion_from_euler(req.roll, req.pitch, req.yaw)
        self.get_logger().info("Corresponding quaternion x: %f, y: %f, z: %f, w: %f" % (res.x, res.y, res.z, res.w))  # Registra el cuaternión resultante (valores de res.*)
        return res
    
    def quaternionToEulerCallback(self, req, res):  # Callback del servicio: recibe la petición (req) y escribe la respuesta (res)
        self.get_logger().info("Requested to convert quaternion x: %f, y: %f, z %f, w: %f" % (req.x, req.y, req.z, req.w))
        (res.roll, res.pitch, res.yaw) = euler_from_quaternion((req.x, req.y, req.z, req.w))
        self.get_logger().info("Corresponding euler angles roll: %f, pitch: %f, yaw: %f " % ( res.roll, res.pitch, res.yaw))
        return res  # Devuelve la respuesta del servicio


def main():  # Función principal (debe estar fuera de la clase)
    rclpy.init()  # Inicializa la comunicación con ROS 2
    angles_converter = AnglesConverter()  # Crea una instancia del nodo
    rclpy.spin(angles_converter)  # Mantiene el nodo ejecutándose para atender solicitudes
    angles_converter.destroy_node()  # Destruye el nodo al finalizar
    rclpy.shutdown()  # Cierra la comunicación con ROS 2


if __name__ == "__main__":  # Punto de entrada cuando se ejecuta el script directamente
    main()  # Llama a la función principal

File: arduinobot_ws/src/arduinobot_utils/src/angle_conversion.cpp
#include <rclcpp/rclcpp.hpp>  // Cabecera principal de ROS 2 en C++ (nodos, logging, etc.)
#include <memory>              // Punteros inteligentes std::shared_ptr, std::make_shared
#include <functional>          // std::bind y std::placeholders
#include <arduinobot_msgs/srv/euler_to_quaternion.hpp>     // Interfaz del servicio EulerToQuaternion (req: roll,pitch,yaw; res: x,y,z,w)
#include <arduinobot_msgs/srv/quaternion_to_euler.hpp>     // Interfaz del servicio QuaternionToEuler (req: x,y,z,w; res: roll,pitch,yaw)
#include <tf2/LinearMath/Quaternion.h>   // tf2::Quaternion
#include <tf2/LinearMath/Matrix3x3.h>    // tf2::Matrix3x3

using namespace std::placeholders;  // Importa _1, _2 para enlazar parámetros en std::bind

// Nodo que ofrece dos servicios para convertir entre Euler <-> Cuaternión
class AnglesConverter : public rclcpp::Node{
    public:
        AnglesConverter() : Node("angles_conversion_service"){  // Constructor: nombra el nodo
            // Servidor: Euler -> Quaternion
            euler_to_quaternion = create_service<arduinobot_msgs::srv::EulerToQuaternion>(
                "euler_to_quaternion",                           // Nombre del servicio en el grafo ROS 2
                std::bind(                                       // Enlaza el callback miembro con this
                    &AnglesConverter::eulerToQuaternionCallback, // Método callback
                    this,                                        // Instancia actual del objeto
                    _1,                                          // Placeholder para Request
                    _2                                           // Placeholder para Response
                )
            );

            // Servidor: Quaternion -> Euler
            quaternion_to_euler = create_service<arduinobot_msgs::srv::QuaternionToEuler>(
                "quaternion_to_euler",                           // Nombre del servicio en el grafo ROS 2
                std::bind(
                    &AnglesConverter::quaternionToEulerCallback, // Método callback
                    this,
                    _1,
                    _2
                )
            );
            
            // Mensaje informativo de arranque
            RCLCPP_INFO(
                rclcpp::get_logger("rclcpp"),
                "Angle conversion services are ready"
            );
        }

    private:
        // Manejadores de servidores de servicio
        rclcpp::Service<arduinobot_msgs::srv::EulerToQuaternion>::SharedPtr euler_to_quaternion;
        rclcpp::Service<arduinobot_msgs::srv::QuaternionToEuler>::SharedPtr quaternion_to_euler;

        // Callback: convierte (roll,pitch,yaw) -> (x,y,z,w)
        void eulerToQuaternionCallback(
            const std::shared_ptr<arduinobot_msgs::srv::EulerToQuaternion::Request> req,  // Petición con ángulos de Euler
            const std::shared_ptr<arduinobot_msgs::srv::EulerToQuaternion::Response> res  // Respuesta con cuaternión
        ){
            RCLCPP_INFO_STREAM(
                rclcpp::get_logger("rclcpp"),
                "Request to convert euler angles roll: " << req->roll
                << ", pitch: " << req->pitch
                << ", yaw: "   << req->yaw
                << " into a quaternion"
            );

            tf2::Quaternion q;
            q.setRPY(req->roll, req->pitch, req->yaw);  // Construye el cuaternión a partir de Euler (RPY)

            res->x = q.getX();
            res->y = q.getY();
            res->z = q.getZ();
            res->w = q.getW();

            RCLCPP_INFO_STREAM(
                rclcpp::get_logger("rclcpp"),
                "Corresponding quaternion x: " << res->x
                << ", y: " << res->y
                << ", z: " << res->z
                << ", w: " << res->w
            );
        }

        // Callback: convierte (x,y,z,w) -> (roll,pitch,yaw)
        void quaternionToEulerCallback(
            const std::shared_ptr<arduinobot_msgs::srv::QuaternionToEuler::Request> req,  // Petición con cuaternión
            const std::shared_ptr<arduinobot_msgs::srv::QuaternionToEuler::Response> res  // Respuesta con ángulos de Euler
        ){
            RCLCPP_INFO_STREAM(
                rclcpp::get_logger("rclcpp"),
                "Request to convert quaternions x: " << req->x
                << ", y: " << req->y
                << " , z: " << req->z
                << " , w: " << req->w
                << " into euler angles"
            );

            tf2::Quaternion q(req->x, req->y, req->z, req->w);  // Construye el cuaternión con los 4 coeficientes
            tf2::Matrix3x3 m(q);                                // Matriz de rotación a partir del cuaternión
            m.getRPY(res->roll, res->pitch, res->yaw);          // Extrae roll, pitch, yaw

            RCLCPP_INFO_STREAM(
                rclcpp::get_logger("rclcpp"),
                "Corresponding Euler Angles roll: " << res->roll
                << ", pitch: " << res->pitch
                << ", yaw: "   << res->yaw
            );
        }
};

int main(int argc, char* argv[]){                 // Función principal
    rclcpp::init(argc, argv);                     // Inicializa ROS 2 con argumentos de línea de comandos
    auto node = std::make_shared<AnglesConverter>();  // Crea la instancia del nodo servidor
    rclcpp::spin(node);                           // Mantiene el nodo ejecutándose para atender solicitudes
    rclcpp::shutdown();                           // Cierra y limpia recursos de ROS 2
    return 0;                                     // Código de salida correcto
}